# Analyze Internet Use with Python

- [View Solution Notebook](./solutions.html)
- [View Project Page](https://www.codecademy.com/projects/practice/analyze-internet-use-with-python)

## Task Group 1 - Import and Inspect

### Task 1

Import the CSV file `internet.csv` and assign it to the variable `internet`. Preview the first few rows.

In [1]:
import pandas as pd
internet = pd.read_csv('internet.csv')
internet.head(10)

entity code  year  internet_users_per_100
0  Afghanistan  AFG  1990                0.000000
1  Afghanistan  AFG  1991                0.000000
2  Afghanistan  AFG  1992                0.000000
3  Afghanistan  AFG  1993                0.000000
4  Afghanistan  AFG  1994                0.000000
5  Afghanistan  AFG  1995                0.000000
6  Afghanistan  AFG  2001                0.004723
7  Afghanistan  AFG  2002                0.004561
8  Afghanistan  AFG  2003                0.087891
9  Afghanistan  AFG  2004                0.105809

<details>
    <summary style="display:list-item; font-size:16px; color:blue;"><i>What is the structure of this dataset? Toggle to check!</i></summary>

Each row of the dataset records the percent of a country's population that uses the internet. According to Our World in Data, a person is defined as an `internet user` by the International Telecommunication Union if they accessed the internet from any location in the last three months.

The columns of the dataset are
- `entity`, the name of the country, region, or income bracket
- `code`, the three-letter country code
- `year`, the year
- `internet_users_per_100`, the number of internet users for every 100 people in the entity's population

Note that we have already trimmed the source dataset to 1990-2019.
</details>

 ### Task 2

The column `internet_users_per_100` is well-named, but let's shorten it.

Clean the dataset by updating the column name `internet_users_per_100` to `percent_online`. Print out `.info()` after doing so. Are there any incorrect data types or missing data issues to be concerned about before diving in?

In [12]:
internet = internet.rename(columns={'internet_users_per_100' : 'percent_online'})
internet.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6367 entries, 0 to 6366
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   entity          6367 non-null   object 
 1   code            6367 non-null   object 
 2   year            6367 non-null   int64  
 3   percent_online  6367 non-null   float64
dtypes: float64(1), int64(1), object(2)
memory usage: 199.1+ KB


## Task Group 2 - Years to Reach Mainstream Use

### Task 3

Let's analyze how long it takes for internet use to reach the majority of people in an `entity`.

Write a function that takes a `row` as input and classifies it by `percent_online`:
- `0` should be classified as `none`
- greater than 0 and under 25 should be classified as `few`
- 25-50 should be classified as `some`
- over 50 should be classified as `most`


In [22]:
def internet_classification():
        if row['percent_online'] == 0:
            return 'none'
        elif row['percent_online'] < 25:
            return 'few'
        elif row['percent_online'] < 50:
            return 'some'
        else:
            return 'most'

### Task 4

Create a new column named `amount` on `internet` by applying the function from Task 3 to each row.

In [25]:
internet['amount'] = internet.apply(internet_classification, axis=1)
internet.head()

entity code  year  percent_online amount
0  Afghanistan  AFG  1990             0.0   none
1  Afghanistan  AFG  1991             0.0   none
2  Afghanistan  AFG  1992             0.0   none
3  Afghanistan  AFG  1993             0.0   none
4  Afghanistan  AFG  1994             0.0   none

### Task 5

Let's figure out the first year in our data that each entity reached each level of internet use.

Pivot `internet` using
- `year` as the values
- `entity` and `code` as the index
- `amount` as the column
- `min` as the aggfunc

Reset the index and save the result as `years`.

In [29]:
internet_pivot = internet.pivot_table(index=['entity', 'code'], 
                                      columns='amount', 
                                      values='year', 
                                      aggfunc='min').reset_index()
internet_pivot.head(50)

amount                    entity code     few    most    none    some
0                    Afghanistan  AFG  2001.0     NaN  1990.0     NaN
1                        Albania  ALB  1995.0  2013.0  1990.0  2009.0
2                        Algeria  DZA  1994.0  2019.0  1990.0  2014.0
3                 American Samoa  ASM     NaN     NaN  1990.0     NaN
4                        Andorra  AND  1996.0  2007.0  1990.0  2004.0
5                         Angola  AGO  1996.0     NaN  1990.0  2017.0
6            Antigua and Barbuda  ATG  1995.0  2011.0  1990.0  2005.0
7                      Argentina  ARG  1992.0  2011.0  1990.0  2007.0
8                        Armenia  ARM  1994.0  2014.0  1990.0  2010.0
9                          Aruba  ABW  1996.0  2008.0  1990.0  2005.0
10                     Australia  AUS  1990.0  2001.0     NaN  1998.0
11                       Austria  AUT  1990.0  2004.0     NaN  2000.0
12                    Azerbaijan  AZE  1994.0  2011.0  1990.0  2009.0
13                       Bahamas  BHS  1995.0  2011.0  1990.0  2005.0
14                       Bahrain  BHR  1995.0  2008.0  1990.0  2006.0
15                    Bangladesh  BGD  1997.0     NaN  1990.0     NaN
16                      Barbados  BRB  1995.0  2005.0  1990.0  2002.0
17                       Belarus  BLR  1994.0  2013.0  1990.0  2009.0
18                       Belgium  BEL  1990.0  2004.0     NaN  2000.0
19                        Belize  BLZ  1995.0  2019.0  1990.0  2008.0
20                         Benin  BEN  1996.0     NaN  1990.0     NaN
21                       Bermuda  BMU  1995.0  2002.0  1990.0  1998.0
22                        Bhutan  BTN  1999.0  2019.0  1990.0  2014.0
23                       Bolivia  BOL  1995.0     NaN  1990.0  2011.0
24        Bosnia and Herzegovina  BIH  1996.0  2015.0  1990.0  2006.0
25                      Botswana  BWA  1995.0  2018.0  1990.0  2013.0
26                        Brazil  BRA  1991.0  2013.0  1990.0  2006.0
27        British Virgin Islands  VGB  2002.0  2016.0  1990.0  2009.0
28                        Brunei  BRN  1995.0  2010.0  1990.0  2004.0
29                      Bulgaria  BGR  1993.0  2012.0  1990.0  2006.0
30                  Burkina Faso  BFA  1996.0     NaN  1990.0     NaN
31                       Burundi  BDI  1996.0     NaN  1990.0     NaN
32                      Cambodia  KHM  1997.0     NaN  1990.0  2016.0
33                      Cameroon  CMR  1997.0     NaN  1990.0  2018.0
34                        Canada  CAN  1990.0  2000.0     NaN  1999.0
35                    Cape Verde  CPV  1997.0  2016.0  1990.0  2010.0
36                Cayman Islands  CYM  1995.0  2007.0  1990.0  2005.0
37      Central African Republic  CAF  1996.0     NaN  1990.0     NaN
38                          Chad  TCD  1997.0     NaN  1990.0     NaN
39                         Chile  CHL  1992.0  2011.0  1990.0  2003.0
40                         China  CHN  1993.0  2015.0  1990.0  2009.0
41                      Colombia  COL  1994.0  2013.0  1990.0  2008.0
42                       Comoros  COM  1998.0     NaN  1990.0     NaN
43                         Congo  COG  1996.0     NaN  1990.0     NaN
44                    Costa Rica  CRI  1992.0  2014.0  1990.0  2006.0
45                 Cote d'Ivoire  CIV  1995.0     NaN  1990.0  2016.0
46                       Croatia  HRV  1993.0  2009.0  1990.0  2004.0
47                          Cuba  CUB  1995.0  2017.0  1990.0  2013.0
48                       Curacao  CUW     NaN  2016.0     NaN     NaN
49                        Cyprus  CYP  1992.0  2010.0  1990.0  2002.0

### Task 6

Albania took from 1995-2009 to reach 25%+ of the country online, corresponding to 14 years. 

By contrast, it only took from 2009-2013, 4 years, to reach over 50%. That's over 3 times faster!

Let's see if this phenomenon holds in general.

Create two new columns:
- `few2some` should be the number of years between `few` and `some`
- `some2most` should be the number of years between `some` and `most`

In [31]:
internet_pivot['few2some'] = internet_pivot['some'] - internet_pivot['few']
internet_pivot['some2most'] = internet_pivot['most'] - internet_pivot['some']

### Task 7

Print out the average of `few2some` and the average of `some2most`.

In [32]:
print(internet_pivot['few2some'].mean())
print(internet_pivot['some2most'].mean())
internet_pivot.head(50)

14.065573770491802
4.929078014184397


amount                    entity code     few    most    none    some  \
0                    Afghanistan  AFG  2001.0     NaN  1990.0     NaN   
1                        Albania  ALB  1995.0  2013.0  1990.0  2009.0   
2                        Algeria  DZA  1994.0  2019.0  1990.0  2014.0   
3                 American Samoa  ASM     NaN     NaN  1990.0     NaN   
4                        Andorra  AND  1996.0  2007.0  1990.0  2004.0   
5                         Angola  AGO  1996.0     NaN  1990.0  2017.0   
6            Antigua and Barbuda  ATG  1995.0  2011.0  1990.0  2005.0   
7                      Argentina  ARG  1992.0  2011.0  1990.0  2007.0   
8                        Armenia  ARM  1994.0  2014.0  1990.0  2010.0   
9                          Aruba  ABW  1996.0  2008.0  1990.0  2005.0   
10                     Australia  AUS  1990.0  2001.0     NaN  1998.0   
11                       Austria  AUT  1990.0  2004.0     NaN  2000.0   
12                    Azerbaijan  AZE  1994.0  2011.0  1990.0  2009.0   
13                       Bahamas  BHS  1995.0  2011.0  1990.0  2005.0   
14                       Bahrain  BHR  1995.0  2008.0  1990.0  2006.0   
15                    Bangladesh  BGD  1997.0     NaN  1990.0     NaN   
16                      Barbados  BRB  1995.0  2005.0  1990.0  2002.0   
17                       Belarus  BLR  1994.0  2013.0  1990.0  2009.0   
18                       Belgium  BEL  1990.0  2004.0     NaN  2000.0   
19                        Belize  BLZ  1995.0  2019.0  1990.0  2008.0   
20                         Benin  BEN  1996.0     NaN  1990.0     NaN   
21                       Bermuda  BMU  1995.0  2002.0  1990.0  1998.0   
22                        Bhutan  BTN  1999.0  2019.0  1990.0  2014.0   
23                       Bolivia  BOL  1995.0     NaN  1990.0  2011.0   
24        Bosnia and Herzegovina  BIH  1996.0  2015.0  1990.0  2006.0   
25                      Botswana  BWA  1995.0  2018.0  1990.0  2013.0   
26                        Brazil  BRA  1991.0  2013.0  1990.0  2006.0   
27        British Virgin Islands  VGB  2002.0  2016.0  1990.0  2009.0   
28                        Brunei  BRN  1995.0  2010.0  1990.0  2004.0   
29                      Bulgaria  BGR  1993.0  2012.0  1990.0  2006.0   
30                  Burkina Faso  BFA  1996.0     NaN  1990.0     NaN   
31                       Burundi  BDI  1996.0     NaN  1990.0     NaN   
32                      Cambodia  KHM  1997.0     NaN  1990.0  2016.0   
33                      Cameroon  CMR  1997.0     NaN  1990.0  2018.0   
34                        Canada  CAN  1990.0  2000.0     NaN  1999.0   
35                    Cape Verde  CPV  1997.0  2016.0  1990.0  2010.0   
36                Cayman Islands  CYM  1995.0  2007.0  1990.0  2005.0   
37      Central African Republic  CAF  1996.0     NaN  1990.0     NaN   
38                          Chad  TCD  1997.0     NaN  1990.0     NaN   
39                         Chile  CHL  1992.0  2011.0  1990.0  2003.0   
40                         China  CHN  1993.0  2015.0  1990.0  2009.0   
41                      Colombia  COL  1994.0  2013.0  1990.0  2008.0   
42                       Comoros  COM  1998.0     NaN  1990.0     NaN   
43                         Congo  COG  1996.0     NaN  1990.0     NaN   
44                    Costa Rica  CRI  1992.0  2014.0  1990.0  2006.0   
45                 Cote d'Ivoire  CIV  1995.0     NaN  1990.0  2016.0   
46                       Croatia  HRV  1993.0  2009.0  1990.0  2004.0   
47                          Cuba  CUB  1995.0  2017.0  1990.0  2013.0   
48                       Curacao  CUW     NaN  2016.0     NaN     NaN   
49                        Cyprus  CYP  1992.0  2010.0  1990.0  2002.0   

amount  few2some  some2most  
0            NaN        NaN  
1           14.0        4.0  
2           20.0        5.0  
3            NaN        NaN  
4            8.0        3.0  
5           21.0        NaN  
6           10.0        6.0  
7           15.0        4.0  
8     

<details>
    <summary style="display:list-item; font-size:16px; color:blue;"><i>What did we learn about the spread of internet use? Toggle to check!</i></summary>

Of the countries in the dataset that have passed the 25% threshold, it takes on average `14` years to go from non-zero internet use to over `25%`. But then, for countries that pass 25% it takes only `5` more years on average to reach a majority of the country.
    
There are some ways that this analysis could be improved. If a country "passes" 25% by jumping suddenly to 45%, for example, it wouldn't be surprising that it takes less time to pass 50%. Think about how you could refine our analysis to take this into account!
</details>

### Task 8

This dataset includes some rows that correspond to broader geographic regions and income brackets, not just countries. These are labeled `CAT` in the `code` column.

Display `years` filtered to rows where `code` is `CAT`. Sort by `few2some`, ascending.

In [35]:
internet_pivot[internet_pivot['code'] == 'CAT'].sort_values('few2some', ascending=True)

amount                        entity code     few    most    none    some  \
151                    North America  CAT  1990.0  2001.0     NaN  1998.0   
87                       High income  CAT  1990.0  2004.0     NaN  2000.0   
66                    European Union  CAT  1990.0  2006.0     NaN  2002.0   
65           Europe and Central Asia  CAT  1990.0  2009.0     NaN  2003.0   
109      Latin America and Caribbean  CAT  1991.0  2015.0  1990.0  2008.0   
56             East Asia and Pacific  CAT  1990.0  2016.0     NaN  2008.0   
218              Upper middle income  CAT  1991.0  2015.0  1990.0  2009.0   
133     Middle East and North Africa  CAT  1990.0  2016.0     NaN  2010.0   
117            Low and middle income  CAT  1991.0     NaN  1990.0  2012.0   
134                    Middle income  CAT  1991.0     NaN  1990.0  2012.0   
119              Lower middle income  CAT  1992.0     NaN  1990.0  2017.0   
189                       South Asia  CAT  1992.0     NaN  1990.0  2019.0   
194               Sub-Saharan Africa  CAT  1991.0     NaN  1990.0  2019.0   
118                       Low income  CAT  1994.0     NaN  1990.0     NaN   

amount  few2some  some2most  
151          8.0        3.0  
87          10.0        4.0  
66          12.0        4.0  
65          13.0        6.0  
109         17.0        7.0  
56          18.0        8.0  
218         18.0        6.0  
133         20.0        6.0  
117         21.0        NaN  
134         21.0        NaN  
119         25.0        NaN  
189         27.0        NaN  
194         28.0        NaN  
118          NaN        NaN

<details>
    <summary style="display:list-item; font-size:16px; color:blue;"><i>What did we learn about the spread of internet use? Toggle to check!</i></summary>

Perhaps unsurprisingly, we see a pretty strong disparity in internet use. The global north and higher income brackets tend to have reached the `most` category fairly quickly. At least in this data, low/middle income brackets and southern regions like South Asia have yet to reach the `most` category, and take the longest to go from `few` to `some`.
</details>

## Task Group 3 - Growth by Decade

### Task 9

Let's return to our original `internet` dataset and investigate how quickly internet use has increased over the decades. 

Note that we don't have full data for all countries. For example, for Nauru we only have one year of data in the 2000s and two years of data in the 2010s. For this project, we'll use the rough approximation of analyzing growth over the years we do have, but to make any strong scientific conclusions we'd want to refine this analysis.

Most of the missing years are in the 1990s, so start by filtering the DataFrame down to only years past 1999.

In [37]:
sub_internet = internet[internet['year'] > 1999]
sub_internet.head(50)

entity code  year  percent_online amount
6   Afghanistan  AFG  2001        0.004723    few
7   Afghanistan  AFG  2002        0.004561    few
8   Afghanistan  AFG  2003        0.087891    few
9   Afghanistan  AFG  2004        0.105809    few
10  Afghanistan  AFG  2005        1.224148    few
11  Afghanistan  AFG  2006        2.107124    few
12  Afghanistan  AFG  2007        1.900000    few
13  Afghanistan  AFG  2008        1.840000    few
14  Afghanistan  AFG  2009        3.550000    few
15  Afghanistan  AFG  2010        4.000000    few
16  Afghanistan  AFG  2011        5.000000    few
17  Afghanistan  AFG  2012        5.454545    few
18  Afghanistan  AFG  2013        5.900000    few
19  Afghanistan  AFG  2014        7.000000    few
20  Afghanistan  AFG  2015        8.260000    few
21  Afghanistan  AFG  2016       11.000000    few
22  Afghanistan  AFG  2017       13.500000    few
23  Afghanistan  AFG  2018       16.799999    few
24  Afghanistan  AFG  2019       17.600000    few
35      Albania  ALB  2000        0.114097    few
36      Albania  ALB  2001        0.325798    few
37      Albania  ALB  2002        0.390081    few
38      Albania  ALB  2003        0.971900    few
39      Albania  ALB  2004        2.420388    few
40      Albania  ALB  2005        6.043891    few
41      Albania  ALB  2006        9.609991    few
42      Albania  ALB  2007       15.036116    few
43      Albania  ALB  2008       23.860001    few
44      Albania  ALB  2009       41.200001   some
45      Albania  ALB  2010       45.000000   some
46      Albania  ALB  2011       47.000000   some
47      Albania  ALB  2012       49.400002   some
48      Albania  ALB  2013       51.799999   most
49      Albania  ALB  2014       54.299999   most
50      Albania  ALB  2015       56.900002   most
51      Albania  ALB  2016       59.599998   most
52      Albania  ALB  2017       62.400002   most
53      Albania  ALB  2018       65.400002   most
54      Albania  ALB  2019       68.550392   most
65      Algeria  DZA  2000        0.491706    few
66      Algeria  DZA  2001        0.646114    few
67      Algeria  DZA  2002        1.591641    few
68      Algeria  DZA  2003        2.195360    few
69      Algeria  DZA  2004        4.634475    few
70      Algeria  DZA  2005        5.843942    few
71      Algeria  DZA  2006        7.375985    few
72      Algeria  DZA  2007        9.451191    few
73      Algeria  DZA  2008       10.180000    few
74      Algeria  DZA  2009       11.230000    few
75      Algeria  DZA  2010       12.500000    few

### Task 10

Now let's add decade information!

Create a function named `decader` that takes a `row` as input and outputs the decade corresponding to `row['year']`.

The format for the decade should be as follows:

- the years `2000-2009` inclusive should have decade `2000s`
- the years `2010-2019` inclusive should have decade `2010s`

In [54]:
def decader(row):
    decade = str(row['year'])
    decade = decade[0:3] + "0s"
    return decade

### Task 11

Use `.apply()` to create a new column `decade` that stores decade information for each row.

In [56]:
sub_internet['decade'] = sub_internet.apply(decader, axis=1)
sub_internet.head()

entity code  year  percent_online amount decade
6   Afghanistan  AFG  2001        0.004723    few  2000s
7   Afghanistan  AFG  2002        0.004561    few  2000s
8   Afghanistan  AFG  2003        0.087891    few  2000s
9   Afghanistan  AFG  2004        0.105809    few  2000s
10  Afghanistan  AFG  2005        1.224148    few  2000s

### Task 12

Now, let's write a function to calculate how much internet use has changed over a sequence of years.

Write a function `change` that takes a DataFrame `column` as input.

If the `len(column)` is `1` (that is, the column has only one entry), return that entry (`.iloc[0]`)

Otherwise, return the difference between the last entry (`.iloc[-1]`) and the first entry (`.iloc[0]`).

In [64]:
def change(column):
    if len(column) == 1:
        return column.iloc[0]
    else:
        return column.iloc[-1] - column.iloc[0]

### Task 13

In order for `change` to compute the change from the start of a decade to its end, we need the years to be in order, so that the last entry of the column is the latest year of the decade in the dataset.

Sort `internet` by `year`, ascending.

In [65]:
internet = internet.sort_values('year', ascending=True)
internet.head()

entity code  year  percent_online amount
0             Afghanistan  AFG  1990             0.0   none
1314                 Cuba  CUB  1990             0.0   none
3131              Lebanon  LBN  1990             0.0   none
6073  Upper middle income  CAT  1990             0.0   none
4427                Palau  PLW  1990             0.0   none

### Task 14

Let's apply `change` to calculate how much the internet has grown over each decade. At the same time, let's calculate the first and last year of data for each country, so we can know how many years it took for that change to occur.

Group `internet` by `entity` and `decade`. Using `.agg`:

- apply `change` to `percent_online`
- apply `['min','max']` to `year`

Assign the result to the variable `decade_growth`, and reset the index.

In [68]:
internet_change = sub_internet.groupby(['entity', 'decade']).agg({'percent_online': change,
                                                             'year': ['min', 'max']}).reset_index()
internet_change

entity decade percent_online  year      
                                change   min   max
0    Afghanistan  2000s       3.545277  2001  2009
1    Afghanistan  2010s      13.600000  2010  2019
2        Albania  2000s      41.085903  2000  2009
3        Albania  2010s      23.550392  2010  2019
4        Algeria  2000s      10.738294  2000  2009
..           ...    ...            ...   ...   ...
441        Yemen  2010s      14.368355  2010  2017
442       Zambia  2000s       2.308928  2000  2009
443       Zambia  2010s      13.799999  2010  2019
444     Zimbabwe  2000s       3.598566  2000  2009
445     Zimbabwe  2010s      18.700000  2010  2019

[446 rows x 5 columns]

### Task 15

Let's flatten the columns of this new DataFrame.

Reassign `.columns` using the names `['entity','decade','change','min','max']`

In [70]:
internet_change.columns = ['entity','decade','change','min','max']
internet_change.head()

entity decade     change   min   max
0  Afghanistan  2000s   3.545277  2001  2009
1  Afghanistan  2010s  13.600000  2010  2019
2      Albania  2000s  41.085903  2000  2009
3      Albania  2010s  23.550392  2010  2019
4      Algeria  2000s  10.738294  2000  2009

### Task 16

Let's calculate yearly growth.

Add a new column `annual` by calculating `decade_growth['change']` divided by `decade_growth['max'] - decade_growth['min']`

In [74]:
internet_change['annual_growth%'] = internet_change['change'] / (internet_change['max'] - internet_change['min'])
internet_change.head()

entity decade     change   min   max  annual_growth  annual_growth%
0  Afghanistan  2000s   3.545277  2001  2009       0.443160        0.443160
1  Afghanistan  2010s  13.600000  2010  2019       1.511111        1.511111
2      Albania  2000s  41.085903  2000  2009       4.565100        4.565100
3      Albania  2010s  23.550392  2010  2019       2.616710        2.616710
4      Algeria  2000s  10.738294  2000  2009       1.193144        1.193144

### Task 17

Let's make this table a bit more human-readable by converting to wide format.

Pivot `decade_growth` using
- `annual` as the values
- `entity` as the index
- `decade` as the columns

Assign the result back to `decade_growth`.

In [77]:
# internet_change = internet_change.pivot_table(index='entity', columns='decade', values='annual_growth')
internet_change

decade          2000s     2010s
entity                         
Afghanistan  0.443160  1.511111
Albania      4.565100  2.616710
Algeria      1.193144  5.044445
Andorra      7.554574  1.509638
Angola       0.243884  3.288889
...               ...       ...
Vietnam      2.921750  4.223509
World        2.103932  2.786652
Yemen        1.097500  2.052622
Zambia       0.256548  1.533333
Zimbabwe     0.399841  2.077778

[225 rows x 2 columns]

### Task 18

Create a new column `ratio` that takes `2010s` and divides by `2000s`.

In [79]:
internet_change['ratio'] = internet_change['2010s']/internet_change['2000s']
internet_change['average'] = ratio
internet_change


decade          2000s     2010s      ratio
entity                                    
Afghanistan  0.443160  1.511111   3.409857
Albania      4.565100  2.616710   0.573199
Algeria      1.193144  5.044445   4.227860
Andorra      7.554574  1.509638   0.199831
Angola       0.243884  3.288889  13.485475
...               ...       ...        ...
Vietnam      2.921750  4.223509   1.445541
World        2.103932  2.786652   1.324497
Yemen        1.097500  2.052622   1.870271
Zambia       0.256548  1.533333   5.976798
Zimbabwe     0.399841  2.077778   5.196514

[225 rows x 3 columns]

### Task 19

Print out descriptive statistics of the new `ratio` column. What do you notice?

In [80]:
internet_change['ratio'].describe()

count    220.000000
mean       5.119733
std       14.289239
min        0.000000
25%        0.718154
50%        1.579371
75%        4.226106
max      140.184335
Name: ratio, dtype: float64

<details>
    <summary style="display:list-item; font-size:16px; color:blue;"><i>What did we learn about internet use? Toggle to check!</i></summary>

On average, annual growth in the 2010s was 5 times as large as annual growth in the 2000s.
    
This average is being pulled up by some enormous outliers. The median is `1.5`, so half the entities in the dataset had over 1.5x more annual growth in the 2010s than the 2000s.
    
Interestingly, the 25th percentile is `.71`, so over 25% of countries experienced less annual growth in the 2010s than in the 2000s! This likely makes sense for countries like the USA, which had already reached a majority of the population by 2010. What other factors might cause a country to end up with slower internet growth in the 2010s than in the 2000s?
    
Again, for some countries we may be missing some decade information. We encourage you dig deeper into this dataset if you want to refine our analysis!
   
</details>

There is so much more to discover in this dataset. Feel free to add more cells below to extend the analysis, and happy coding!